Feature/3301/refresh individual version

agduncan94 opened the pull request:
Can refresh individual versions (they don't need to already exist on Dockstore). Can also refresh legacy versions in a migrated GH app workflow

https://github.com/dockstore/dockstore/issues/3301

denis-yuen commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Coverage increased per file
===========================
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ObsoleteUrlFactory.java  54
+ dockstore-webservice/src/main/java/io/swagger/api/impl/ToolsApiServiceImpl.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java  29
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/EntryStarredSerializer.java  100
+ dockstore-common/src/main/java/io/dockstore/common/EntryType.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EventResource.java  57
+ dockstore-common/src/main/java/io/dockstore/common/EntryUpdateTime.java  55
+ dockstore-webservice/src/main/java/io/swagger/model/ToolVersionV1.java  52
+ dockstore-webservice/src/main/java/io/swagger/model/ToolDockerfile.java  23
+ dockstore-webservice/src/main/java/io/swagger/api/impl/MetadataApiServiceImpl.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Alias.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/TokenDAO.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/MetadataResourceHelper.java  58
+ dockstore-common/src/main/java/io/dockstore/common/Utilities.java  84
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/ResourceUtilities.java  66
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/PersistenceExceptionMapper.java  100
+ dockstore-common/src/main/java/io/dockstore/common/yaml/AbstractYamlService.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceFactory.java  100
+ dockstore-common/src/main/java/io/dockstore/common/Repository.java  56
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java  77
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/ServiceEntryDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java  69
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Version.java  88
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Label.java  50
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/FileFormatHelper.java  87
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AliasableResourceInterface.java  71
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Token.java  47
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Checksum.java  80
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java  44
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Image.java  60
+ dockstore-webservice/src/main/java/io/swagger/model/ToolV1.java  60
+ dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYamlHelper.java  72
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java  90
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/dockerhub/DockerHubTag.java  15
+ dockstore-webservice/src/main/java/io/openapi/api/impl/ToolsApiServiceImpl.java  80
+ dockstore-common/src/main/java/io/dockstore/common/Registry.java  94
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/ToolMode.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/DescriptionSource.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/BioWorkflowDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/BitBucketSourceCodeRepo.java  17
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerFactory.java  48
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/ToolPath.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/StateManagerMode.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/NotificationResource.java  62
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/HostedWorkflowResource.java  90
+ dockstore-webservice/src/main/java/io/swagger/api/ToolClassesApiV1.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ManualRegistry.java  66
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/LabelDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/EntryVersionHelper.java  65
+ dockstore-common/src/main/java/io/dockstore/common/VersionTypeValidation.java  60
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSEntry.java  79
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/WorkflowMode.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/WDLHandler.java  66
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ConfigHelper.java  55
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ImageRegistryFactory.java  88
+ dockstore-common/src/main/java/io/dockstore/common/OrganizationUpdateTime.java  57
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Organization.java  86
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java  92
+ dockstore-common/src/main/java/io/dockstore/common/yaml/Dockstore12Validator.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/VersionMetadata.java  25
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java  8
+ dockstore-webservice/src/main/java/io/dockstore/webservice/doi/DOIGeneratorFactory.java  66
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/EntryLabelHelper.java  86
+ dockstore-webservice/src/main/java/io/dockstore/webservice/api/PublishRequest.java  50
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/ServicePrototypePlugin.java  38
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceCodeRepoInterface.java  83
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/FileFormatDAO.java  85
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/TRSListener.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AliasResource.java  31
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Workflow.java  93
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Notification.java  57
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/UserDAO.java  61
+ dockstore-webservice/src/main/java/io/openapi/model/ExtendedFileWrapper.java  41
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/ServiceDAO.java  80
+ dockstore-webservice/src/main/java/io/swagger/api/ToolsApiV1.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/SimpleAuthenticator.java  61
+ dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/Role.java  87
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/tooltester/ToolTesterLogType.java  100
+ dockstore-webservice/src/main/java/io/openapi/api/impl/ToolClassesApiServiceImpl.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/FileDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/ToolTesterResource.java  16
+ dockstore-webservice/src/main/java/io/swagger/model/ToolDescriptor.java  44
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowVersionDAO.java  66
+ dockstore-common/src/main/java/io/dockstore/common/SourceControl.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/SimpleAuthorizer.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/ElasticListener.java  89
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/TagDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/TOSVersion.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSWriter.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Tool.java  87
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/HostedToolResource.java  90
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Collection.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/MetadataResource.java  76
+ dockstore-webservice/src/main/java/io/swagger/api/MetadataApiV1.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/URIHelper.java  55
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/PublicStateManager.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceControlConverter.java  77
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/ElasticSearchHealthCheck.java  25
+ dockstore-webservice/src/main/java/io/swagger/model/ToolTestsV1.java  25
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/WorkflowPath.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/ChecksumConverter.java  66
+ dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguageSubclass.java  93
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguagePluginHandler.java  35
+ dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/InMemoryPermissionsImpl.java  69
+ dockstore-common/src/main/java/io/dockstore/common/yaml/YamlWorkflow.java  92
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsExtendedApiService.java  100
+ dockstore-common/src/main/java/io/dockstore/common/yaml/Service12.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Tag.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EventSearchType.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/Permission.java  63
+ dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java  93
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java  76
+ dockstore-webservice/src/main/java/io/dockstore/webservice/api/Config.java  100
+ dockstore-webservice/src/main/java/io/swagger/model/ExtendedFileWrapper.java  41
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/AbstractDockstoreDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionResource.java  73
+ dockstore-common/src/main/java/io/dockstore/common/LanguageHandlerHelper.java  80
+ dockstore-webservice/src/main/java/io/swagger/api/impl/ApiV1VersionConverter.java  86
+ dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYaml12.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceCodeRepoFactory.java  73
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Validation.java  59
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java  74
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSFeed.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/api/Limits.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Service.java  42
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java  9
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TemplateHealthCheck.java  42
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/StateListenerInterface.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/FileFormat.java  83
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractHostedEntryResource.java  85
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/BioWorkflow.java  61
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/RSSWorkflowPath.java  100
+ dockstore-webservice/src/main/java/io/swagger/api/impl/ToolsImplCommon.java  78
+ dockstore-webservice/src/main/java/io/dockstore/webservice/api/StarRequest.java  62
+ dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYaml10.java  50
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/RSSToolPath.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java  85
+ dockstore-common/src/main/java/io/dockstore/common/LanguagePluginManager.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSHeader.java  81
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java  79
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsExtendedApi.java  55
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/ToolDAO.java  84
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/EntryLite.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java  80
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubHelper.java  29
+ dockstore-webservice/src/main/java/io/swagger/api/impl/ToolClassesApiServiceImpl.java  75
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ZipSourceFileHelper.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/NotificationDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/StarrableResourceInterface.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java  48
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java  72
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoTagResource.java  37
+ dockstore-webservice/src/main/java/io/dockstore/webservice/CharsetResponseFilter.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/TransactionExceptionMapper.java  91
+ dockstore-webservice/src/main/java/io/swagger/api/impl/ApiV2BetaVersionConverter.java  85
+ dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/PermissionsFactory.java  69
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/User.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/CustomWebApplicationException.java  75
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/TokenTypeConverter.java  77
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/SourceControlResourceInterface.java  62
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GoogleHelper.java  18
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/RSSListener.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/VersionDAO.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java  66
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/TokenType.java  100
+ dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYaml11.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/SharedWorkflows.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceApplication.java  82
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java  87
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/CollectionOrganization.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java  67
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowDAO.java  60
+ dockstore-common/src/main/java/io/dockstore/common/yaml/YamlService11.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EventDAO.java  56
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/WorkflowVersion.java  69
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CacheConfigManager.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/ServiceResource.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/SitemapListener.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/OrganizationDAO.java  75
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/PrivacyPolicyVersion.java  100
+ dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/PermissionsInterface.java  79
+ dockstore-webservice/src/main/java/io/swagger/model/MetadataV1.java  48
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AuthenticatedResourceInterface.java  46
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java  77
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/CollectionDAO.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/OrganizationUser.java  100
         

Complexity increasing per file
==============================
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java  1
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=5238704&bid=17171052)

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:
```diff
@@ -151,20 +151,24 @@ private String getToken(List<Token> tokens, TokenType tokenType) {
-        for (String version : removedVersions) {
-            workflow.removeWorkflowVersion(existingVersionMap.get(version));
+        // delete versions that exist in old workflow but do not exist in newWorkflow (only for whole refresh)
+        if (versionName.isEmpty()) {
```
What happens if versionName is not empty and the version got deleted on GitHub?

agduncan94 replied:
Good call. Looks like the endpoint will do nothing and not throw an error. I'll update it so that it throws an error if the branch does not exist.

agduncan94 replied:
Actually think I misunderstood. Do you mean if the refresh is of a version that doesn't exist on GitHub but exists on Dockstore, it should be deleted on Dockstore?

agduncan94 replied:
After discussing with Denis, we've decided that on refresh of a version, if that version does not exist, throw an error. This is different from the current refresh workflow logic, which would delete a version on Dockstore if it doesn't exist on the Git repo anymore. We decided on this difference because of how we expect the users to use this refresh in the UI. If they are refreshing a version, then likely they expect it to still exist and are not trying to delete it. A delete version button may be present in the future, pending archive/delete/life cycle discussions.

coverbeck replied:
> Actually think I misunderstood. Do you mean if the refresh is of a version that doesn't exist on GitHub but exists on Dockstore, it should be deleted on Dockstore?

Yes, I meant that. :) Although I think I missed the larger context, and was just assuming it was a deleted version in general, as opposed to the specific version you are refreshing. You have it covered with the other ticket in any case.

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:
```diff
@@ -151,20 +151,24 @@ private String getToken(List<Token> tokens, TokenType tokenType) {
+            newWorkflow.getWorkflowVersions().forEach(version -> newVersionMap.put(version.getName(), version));
+            Sets.SetView<String> removedVersions = Sets.difference(existingVersionMap.keySet(), newVersionMap.keySet());
+            for (String version : removedVersions) {
+                workflow.removeWorkflowVersion(existingVersionMap.get(version));
```
Shouldn't remove frozen versions. While database shouldn't let you, I think the whole transaction may fail, so other changes that happened will get lost.

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java:
```diff
@@ -380,61 +380,109 @@ private void logFullWorkflowRefresh(final Workflow workflow) {
+    }
+
+    @GET
+    @Path("/{workflowId}/refresh/{version}")
```
Paging @garyluu 
Does a shared prefix with `@Path("/{workflowId}/refresh")` cause issues with ELK?

garyluu replied:
No, there's a slash, it should be fine.  but as always, the script to generate the patterns should always be ran.

garyluu commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java:
```diff
@@ -380,61 +380,109 @@ private void logFullWorkflowRefresh(final Workflow workflow) {
     @Path("/{workflowId}/refresh")
     @Timed
     @UnitOfWork
+    @Operation(operationId = "refresh", description = "Refresh one particular workflow.", security = @SecurityRequirement(name = "bearer"))
```
use constant for "bearer" here and below

garyluu commented on dockstore-webservice/src/main/resources/openapi3/openapi.yaml:
```diff
@@ -4115,6 +4115,65 @@ paths:
+        required: true
+        schema:
+          type: string
+      requestBody:
```
user not hidden for both endpoints

agduncan94 replied:
How do you hide with openapi?

garyluu replied:
`@Parameter(hidden = true, name = "user", in = ParameterIn.HEADER)`

agduncan94 commented on dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java:
```diff
@@ -473,10 +476,12 @@ public GHRepository getRepository(String repositoryId) {
+     * @param versionName Optional version name to refresh
      * @return WorkflowVersion for the given reference
      */
+    @SuppressWarnings("checkstyle:ParameterNumber")
```
😢

denis-yuen commented on dockstore-integration-testing/src/test/java/io/dockstore/webservice/WebhookIT.java:
```diff
@@ -117,6 +120,30 @@ public void testWorkflowMigration() throws Exception {
             assertEquals("Should not be able to refresh a dockstore.yml workflow.", HttpStatus.SC_BAD_REQUEST, ex.getCode());
         }
 
+        // Should be able to refresh a legacy version
```
Could use a test for refresh version on a frozen version

agduncan94 replied:
Actually this is hard to test, since we would be testing that a frozen version is not removed when the corresponding Git repo branch is deleted. I don't think our current testing framework supports this.

denis-yuen replied:
That would be hard.  How about testing that a frozen version is not updated when the corresponding Git repo branch is newer?

agduncan94 replied:
Wouldn't that also require making a push to the Git repo?

agduncan94 replied:
What if I do a refresh and check the dbupdatetime?

denis-yuen replied:
Was more thinking that you could start the database off with different content, freeze it, do a fake call to the new endpoint, and just verify that the new endpoint respects the frozen status and doesn't update to whats in GitHub

agduncan94 commented:
@denis-yuen I added a frozen test that would just confirm the commit ID would not update on a frozen version. Also, think I found an issue where the commit id wasn't being copied over to the db version so I fixed that.
